File: week3/q06.py

# Pretty bad merge solution
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []

        l = self.levelOrder(root.left)
        r = self.levelOrder(root.right)
        merge = [(l[i] if i < len(l) else [])+(r[i] if i < len(r) else []) for i in range(max(len(l), len(r)))]

        # merge left and right
        return [[root.val]] + merge

# Much better queue based solution
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        if root is None:
            return []

        q = deque([])
        q.append(root)

        res = []
        while len(q) > 0:
            cur_level = []
            for _ in range(len(q)):
                cur_node = q.popleft()
                cur_level.append(cur_node.val)
                if cur_node.left:
                    q.append(cur_node.left)
                if cur_node.right:
                    q.append(cur_node.right)
            if len(cur_level) > 0:
                res.append(cur_level)

        return res

File: week6/q06.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        auto [ret, height] = _rightSideView(root);
        return ret;
    }

    // pair of height and return value
    pair<vector<int>,int> _rightSideView(TreeNode* root) {
        if (root == nullptr) {
            return {{}, 0}; 
        }

        auto [left, left_height] = _rightSideView(root->left);
        auto [right, right_height] = _rightSideView(root->right);
        
        if (left_height > right_height) {
            auto start = left.begin() + right_height;
            auto end = left.end();
            right.insert(right.end(), start, end);
        }
        right.insert(right.begin(), root->val);
        return {right, max(left_height, right_height) + 1};
    }
};

// Better solution
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {

        if (root == nullptr) return {};

        vector<int> res{};
        deque<TreeNode*> next{root};
        while (next.size() > 0) {
            res.push_back(next.back()->val);
            auto next_copy = next;
            next.clear();
            for (auto next_node : next_copy) {
                if (next_node->left != nullptr) next.push_back(next_node->left);
                if (next_node->right != nullptr) next.push_back(next_node->right);
            }
        }

        return res;
    }

};
